docs/csharp/whats-new/csharp-8.md

```diff
@@ -0,0 +1,356 @@
+
+## Asynchronous streams
+
+Starting with C# 8.0, you can create and consume streams asynchronously. A method that returns an asynchronous stream has three properties:
+
+1. It was declared with the `async` modifier.
+1. It returns an <xref:System.Collections.Generic.IAsyncEnumerable%601>.
+1. The method contains `yield return` statements to return successive elements in the asynchronous stream.
+
+Consuming an asynchronous stream requires you to add the `await` keyword before the `foreach` keyword when you enumerate the elements of the stream. Adding the `await` keyword requires the method that enumerates the asynchronous stream to be declared with the `async` modifier and to return a type allowed for an `async` method. Typically that means returning a <xref:System.Threading.Tasks.Task> or <xref:System.Threading.Tasks.Task%601>. It can also be a <xref:System.Threading.Tasks.ValueTask> or <xref:System.Threading.Tasks.ValueTask%601>. A method can both consume and produce an asynchronous stream, which means it would return an <xref:System.Collections.Generic.IAsyncEnumerable%601>. The following code generates a sequence from 1 to 20, waiting 100 ms between generating each number:
+
+```csharp
+public static async System.Collections.Generic.IAsyncEnumerable<int> GenerateSequence()
+{
+    for (int i = 0; i < 20; i++)
+    {
+        await Task.Delay(100);
+        yield return i;
+    }
+}
+```
+
+You would enumerate the sequence using the `await foreach` statement:
+
+```csharp
+await foreach (var number in GenerateSequence())
+{
+    Console.WriteLine(number);
+}
+```
+
+You can try asynchronous streams yourself in our tutorial on [creating and consuming async streams](../tutorials/generate-consume-asynchronous-stream.md).
+
+## Indices and ranges
+
+Ranges and indices provide a succinct syntax for specifying subranges in an array, <xref:System.Span%601>, or <xref:System.ReadOnlySpan%601>.
+
+You can specify an index **from the end**. You specify **from the end** using the `^` operator. You are familiar with `array[2]` meaning the element "2 from the start". Now, `array[^2]` means the element "2 from the end". The index `^0` means "the end", or one past that last element.
+
+You can specify a **range** with the **range operator**: `..`. For example, `0..^0` specifies the entire range of the array: 0 from the start up to, but not including 0 from the end. Either operand may use "from the start" or "from the end". Furthermore, either operand may be omitted. The defaults are `0` for the start index, and `^0` for the end index.
```

Comment:
It might be worth pointing out that the end of a range is exclusive of the index. Much like how random works. The start is inclusive lower bound of the value, but the end is exclusive upper bound of the value. (I'm assuming this is how range works?)

And additionally, it seem the `^` (from end) indicator is also exclusive upper bound-based.

The "2 from start" doesn't make sense to me either. I would make sure to use **element-index** or something to clarify that. When you say `element "2 from the start"` it sounds like you're pulling the element value of 2 out, versus `element index 2 from the start` or `3rd element from the start` would be clearer. Or compare it more directly and indicate the difference with something like:

> `array[2]` means the 3rd element from the start of the array and `array[^2]` means the 2nd element from the end of the array.

And how can you go passed the end of an array with `^0` it seems like that should be invalid? I've not really investigated this stuff so these are my "first look" thoughts :)